session/session.go

```diff
@@ -11,447 +11,243 @@
+				if err != nil {
+					log.Errorf("failed to ping client: %s", err.Error())
+					return
+				}
+				sessIdle = true
+				timerPing.Reset(pongWait)
 			}
-		case <-time.After(time.Until(nextHealthcheckPing)):
-			s.healthcheckPing()
-			nextHealthcheckPing = time.Now().Add(healthcheckInterval)
-		}
-	}
-}
-
-func (s *MenderShellSession) healthcheckPing() {
-	msg := &ws.ProtoMsg{
-		Header: ws.ProtoHdr{
-			Proto:     ws.ProtoTypeShell,
-			MsgType:   wsshell.MessageTypePingShell,
-			SessionID: s.id,
-			Properties: map[string]interface{}{
-				"timeout": int(healthcheckInterval.Seconds() + healthcheckTimeout.Seconds()),
-				"status":  wsshell.ControlMessage,
-			},
-		},
-		Body: nil,
-	}
-	log.Debugf("session %s healthcheck ping", s.id)
-	err := connectionmanager.Write(ws.ProtoTypeShell, msg)
-	if err != nil {
-		log.Debugf("error on write: %s", err.Error())
-	}
-}
-
-func (s *MenderShellSession) HealthcheckPong() {
-	s.pong <- struct{}{}
-}
-
-func (s *MenderShellSession) ShellCommand(m *ws.ProtoMsg) error {
-	s.activeAt = timeNow()
-	data := m.Body
-	commandLine := string(data)
-	n, err := s.writer.Write(data)
-	if err != nil && n != len(data) {
-		err = shell.ErrExecWriteBytesShort
-	}
-	if err != nil {
-		log.Debugf("error: '%s' while running '%s'.", err.Error(), commandLine)
-	} else {
-		log.Debugf("executed: '%s'", commandLine)
-	}
-	return err
-}
-
-func (s *MenderShellSession) ResizeShell(height, width uint16) {
-	shell.ResizeShell(s.pseudoTTY, height, width)
-}
-
-func (s *MenderShellSession) StopShell() (err error) {
-	log.Infof("session %s status:%d stopping shell", s.id, s.status)
-	if s.status != ActiveSession && s.status != HangedSession {
-		return ErrSessionShellNotRunning
-	}
+			continue
 
-	close(s.stop)
-	s.shell.Stop()
-	s.terminal = MenderShellTerminalSettings{}
-	s.status = EmptySession
+		case msg, open = <-sess.msgChan:
+			if !open {
+				return
+			}
+			timerPing.Reset(sess.Config.IdleTimeout)
+			sessIdle = false
+		}
 
-	p, err := os.FindProcess(s.shellPid)
-	if err != nil {
-		log.Errorf("session %s, shell pid %d, find process error: %s", s.id, s.shellPid, err.Error())
-		return err
-	}
-	err = p.Signal(syscall.SIGINT)
-	if err != nil {
-		log.Errorf("session %s, shell pid %d, signal error: %s", s.id, s.shellPid, err.Error())
-		return err
-	}
-	s.pseudoTTY.Close()
+		if msg.Header.Proto == ws.ProtoTypeControl {
+			if sess.HandleControl(msg) {
+				return
+			}
+			continue
+		}
 
-	err = procps.TerminateAndWait(s.shellPid, s.command, 2*time.Second)
-	if err != nil {
-		log.Errorf("session %s, shell pid %d, termination error: %s", s.id, s.shellPid, err.Error())
-		return err
+		handler, ok := sess.handlers[msg.Header.Proto]
+		if !ok {
+			constructor, ok := sess.Routes[msg.Header.Proto]
+			if !ok {
+				sess.Error(msg, false, fmt.Sprintf(
+					"no handler registered for protocol: 0x%04X",
+					msg.Header.Proto,
+				))
+				continue
+			}
+			handler = constructor()
+			defer handler.Close()
```

Comment:
just for my peace of mind, can you confirm that the Close() method will have been called on exit from this function for every handler?

Reply:
The Close function is pushed on the defer stack as soon as the handler is created. So under normal shutdown it should always call the close function. However, I did not consider exceptional cases such as terminating signals etc. Do you think it's necessary to add a signal handler that ensures all resources are deallocated?